### PR TITLE
Adding Jekyll plugin that generates Hierarchical Master Table of Content

### PR DESCRIPTION
**Pull request by nguyenfilip**
@awood  we talked about this - would you like to review please?

### CODE REVIEW

File: _plugins/docs_index.rb [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+      @tokens = tokens
+    end
+
+    #Adds titles of pages to Master Table of Contents and also 
+    #adds headings of pages to Master Table of Contents.
+    # When a page info is extracted from project_pages hash,
+    #the page is removed from project_pages. This way we can track
+    # uncategorized pages
+    def self.addTitlesAndSubsections(sections, project_pages) 
+      for s in sections do
```
Comment (awood) [inline _plugins/docs_index.rb]: `section.each do |s|` is sort of the standard Ruby idiom here.

File: _plugins/docs_index.rb [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        section_name = s["section"]
+        page = project_pages.delete(section_name);
+        if (page == nil)
```
Comment (awood) [inline _plugins/docs_index.rb]: `if page.nil?` is more idiomatic.

File: _plugins/docs_index.rb [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+          raise "Page '"+section_name+".md' that is defined in Master Table of Contents wasn't found!"
```
Comment (awood) [inline _plugins/docs_index.rb]: You can use string interpolation here.

```
raise "Page #{section_name}.md that is defined ..."
```

File: _plugins/docs_index.rb [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        end
+        s['section-title']=page[:title]
+        
+        #Make sure subs array exists
+        if (s["subs"] == nil)
```
Comment (awood) [inline _plugins/docs_index.rb]: `s['subs'] ||= []`

File: _plugins/docs_index.rb [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    def self.buildTocItems(sections, level, number) 
+      sectionNum = 1;
+      result = []
+      for s in sections do
+        if (number == "")
+          secnum = sectionNum.to_s;
+        else
+          secnum = number+"."+sectionNum.to_s;
+        end
+        result << TocItem.new(level, secnum, s['section-title'], s['section'], s.has_key?("subs") && s["subs"].length > 0);
```
Comment (awood) [inline _plugins/docs_index.rb]: `has_key?` has been [deprecated](http://batsov.com/articles/2013/08/21/the-elements-of-style-in-ruby-number-9-hash-number-has-key-and-hash-number-has-value-are-deprecated/) in favor of `key?`.

File: _plugins/docs_index.rb [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+    end
+
+    #Takes enriched sections yaml and creates numbered hierarchy
+    # of sections     
+    def self.buildTocItems(sections, level, number) 
+      sectionNum = 1;
+      result = []
+      for s in sections do
+        if (number == "")
+          secnum = sectionNum.to_s;
```
Comment (awood) [inline _plugins/docs_index.rb]: Semi-colons are not require here and look out of place in Ruby code.

File: _plugins/docs_index.rb [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+      sections_yaml = page["sections"]
+      
+      all_pages = DocsIndexTag.pagesByName(site,page);
+      
+      #After this method is called, all unused pages will stay in all_pages
+      DocsIndexTag.addTitlesAndSubsections(sections_yaml, all_pages);
+      tocitems = DocsIndexTag.buildTocItems(sections_yaml, 1, "");  
+      
+      uncat=[]
+      for k,v  in all_pages
```
Comment (awood) [inline _plugins/docs_index.rb]: `all_pages.each do |k, v|` works

**Comment by nguyenfilip** [comment]
@awood Thanks for all your comments. I have amended the commit and adressed all of them. Also I moved all the pages we have on our site under "developers, usage,design" into the Master Table of Contents under "old_layout" node.

**Comment by nguyenfilip** [comment]
Looks good to me

**Comment by nguyenfilip** [comment]
These changes actually break deployed docs on Openshift. Had to revert it and will create a new PR for this.

I will test it more on http://test-candlepinproject.rhcloud.com/docs/candlepin/ before actually suggesting merge to master.

**Pull request merged**